substrate/runtime/contract/src/gas.rs

```diff
@@ -0,0 +1,132 @@
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate. If not, see <http://www.gnu.org/licenses/>.
+
+use {Trait, Module};
+use runtime_primitives::traits::{As, CheckedMul, CheckedSub, Zero};
+use staking;
+
+#[must_use]
+#[derive(Debug, PartialEq, Eq)]
+pub enum GasMeterResult {
+	Proceed,
+	OutOfGas,
+}
+
+impl GasMeterResult {
+	pub fn is_out_of_gas(&self) -> bool {
+		match *self {
+			GasMeterResult::OutOfGas => true,
+			GasMeterResult::Proceed => false,
+		}
+	}
+}
+
+pub struct GasMeter<T: Trait> {
+	gas_left: T::Gas,
+	gas_price: T::Balance,
+}
+impl<T: Trait> GasMeter<T> {
+	#[cfg(test)]
+	pub fn with_limit(gas_limit: T::Gas, gas_price: T::Balance) -> GasMeter<T> {
+		GasMeter {
+			gas_left: gas_limit,
+			gas_price,
+		}
+	}
+
+	/// Account for used gas.
+	///
+	/// Returns `OutOfGas` if there is not enough gas or addition of the specified
+	/// amount of gas has lead to overflow. On success returns `Proceed`.
+	///
+	/// NOTE that `amount` is always consumed, i.e. if there is not enough gas
+	/// the will be set to zero.
+	pub fn charge(&mut self, amount: T::Gas) -> GasMeterResult {
+		let new_value = match self.gas_left.checked_sub(&amount) {
+			None => None,
+			Some(val) if val.is_zero() => None,
+			Some(val) => Some(val),
+		};
+
+		// We always consume the gas even if there is not enough gas.
+		self.gas_left = new_value.unwrap_or_else(Zero::zero);
+
+		match new_value {
+			Some(_) => GasMeterResult::Proceed,
+			None => GasMeterResult::OutOfGas,
+		}
+	}
+
+	pub fn charge_by_balance(&mut self, amount: T::Balance) -> GasMeterResult {
+		let amount_in_gas: T::Balance = amount / self.gas_price;
+		let amount_in_gas: T::Gas = <T::Gas as As<T::Balance>>::sa(amount_in_gas);
+		self.charge(amount_in_gas)
+	}
+
+	pub fn with_nested<R, F: FnOnce(Option<&mut GasMeter<T>>) -> R>(
+		&mut self,
+		amount: T::Gas,
+		f: F,
+	) -> R {
+		// NOTE that it is ok to allocate all available gas since it still ensured
+		// by `charge` that it doesn't reach zero.
+		if self.gas_left < amount {
+			f(None)
+		} else {
+			self.gas_left = self.gas_left - amount;
+			let mut nested = GasMeter {
+				gas_left: amount,
+				gas_price: self.gas_price,
+			};
+
+			let r = f(Some(&mut nested));
+
+			self.gas_left = self.gas_left + nested.gas_left;
+
+			r
+		}
+	}
+
+	/// Returns how much gas left from the initial budget.
+	pub fn gas_left(&self) -> T::Gas {
+		self.gas_left
+	}
+}
+
+pub fn buy_gas<T: Trait>(
```

Comment:
where is this used? needs a doc comment.

Reply:
ok looks like it's just call/create. should probably be a little more restricted, like `pub(crate)` or something

Reply:
yeah, and it isn't exported. 

There are quite a few places which relies on the fact they are declared in a not public sub-module. E.g.: `pub struct GasMeter`, `pub enum GasMeterResult` or bunch of others in other modules (e.g. `account_db::AccountDb`).

Should I change them as well?